### PR TITLE
Camera selection option changes

### PR DESCRIPTION
**Pull request by iamianmullins**
Changes to camera selection logic as the application  was failing to run
in some cases where no camera option was selected due to incorrect camera config.

Thanks Eric, let me know your thoughts on this one.

### CODE REVIEW

File: src/camera_session.cpp [hunk 1]
```diff
@@ -49,6 +59,13 @@ int CameraSession::init() {
     return 0;
   }
 
+  validateConfig();
+
+  return 0;
+}
+
+int CameraSession::validateConfig() {
```
Comment (ericcurtin) [inline src/camera_session.cpp:67]: Good idea

File: src/camera_session.h [hunk 2]
```diff
@@ -33,6 +33,7 @@ class CameraSession {
   ~CameraSession();
 
   int init();
+  int validateConfig();
```
Comment (ericcurtin) [inline src/camera_session.h:36]: This can be private right?

File: src/twincam.h [hunk 3]
```diff
@@ -5,6 +5,7 @@
 
 struct options {
   unsigned long camera = 0;
+  bool camera_selected = false;
```
Comment (ericcurtin) [inline src/twincam.h:8]: I'd like to keep one variable per option, so if we change:

`unsigned long camera = 0;`

to

`long camera = -1;`

if (camera < 0) // we never explicitly set the camera

Makes sense?

File: src/camera_session.cpp [hunk 4]
```diff
@@ -27,8 +27,18 @@ using namespace libcamera;
 
 CameraSession::CameraSession(const CameraManager* const cm) {
   PRINT_FUNC();
-  if (opts.camera < cm->cameras().size()) {
-    camera_ = cm->cameras()[opts.camera];
+  if (opts.camera_selected){
```
Comment (ericcurtin) [inline src/camera_session.cpp:30]: Needs a git clang-format

File: src/camera_session.h (hunk 2 above)
Comment (iamianmullins) [reply src/camera_session.h:36]: Yeah, we should be fine to make this private

File: src/twincam.h (hunk 3 above)
Comment (iamianmullins) [reply src/twincam.h:8]: Yeah, that makes more sense. I like your solution more. We'll go with that. Thanks

**Review by ericcurtin — APPROVED** [review]
LGTM, there are conflicts to fix, "else { if" and "else if" are the same. I can fix the conflicts if you don't get around to it.

File: src/camera_session.cpp [hunk 5]
```diff
@@ -50,6 +56,13 @@ int CameraSession::init() {
     return 0;
   }
 
+  validateConfig();
```
Comment (ericcurtin) [inline src/camera_session.cpp:59]: We need to narrow this down to one validateConfig, we will validate a second time in some cases, need to be wary of performance.

File: src/camera_session.cpp [hunk 6]
```diff
@@ -76,7 +89,7 @@ int CameraSession::init() {
 
   config_ = std::move(cfg);
 
-  return 0;
+  return true;
```
Comment (ericcurtin) [inline src/camera_session.cpp:92]: We return true, 2 and 0 here. We should either go all int or all bool. 0 for success and two other numbers for the two other cases would be the most standard way.

File: src/camera_session.cpp (hunk 5 above)
Comment (iamianmullins) [reply src/camera_session.cpp]: I agree. I've removed the second validateConfig. Let me know what you think when you get a chance

File: src/camera_session.cpp [hunk 7]
```diff
@@ -29,8 +29,15 @@ using namespace libcamera;
...
-  if (opts.camera < cm->cameras().size()) {
+  if (opts.camera < 0) {
+    for (size_t i = 0; i < cm->cameras().size(); ++i) {
+      camera_ = cm->cameras()[i];
+      if (CameraSession::validateConfig())
+        return;
+    }
+  } else if (opts.camera < (int)cm->cameras().size()) {
     camera_ = cm->cameras()[opts.camera];
+    CameraSession::validateConfig();
```
[the remark this answers is not in the document]
Comment (iamianmullins) [reply src/camera_session.cpp]: Thanks Eric, good point, I've made some changes to address this

File: src/camera_session.cpp [hunk 8]
```diff
@@ -77,7 +89,7 @@ int CameraSession::init() {
 
   config_ = std::move(cfg);
 
-  return 0;
+  return 2;
```
[the remark this answers is not in the document]
Comment (iamianmullins) [reply src/camera_session.cpp]: Yes, my mistake. I have corrected this also.

File: src/camera_session.cpp [hunk 9]
```diff
@@ -65,7 +81,7 @@ int CameraSession::init() {
       break;
 
     case CameraConfiguration::Adjusted:
-      break;
+      return 1;
```
Comment (ericcurtin) [inline src/camera_session.cpp:84]: This should not be return 1, it's a valid case, it just means the config was adjusted

Comment (ericcurtin) [reply src/camera_session.cpp:84]: It should break

File: src/camera_session.cpp [hunk 10]
```diff
@@ -29,8 +29,19 @@ using namespace libcamera;
 
 CameraSession::CameraSession(const CameraManager* const cm) {
   PRINT_FUNC();
-  if (opts.camera < cm->cameras().size()) {
+  if (opts.camera < 0) {
+    for (size_t i = 0; i < cm->cameras().size(); ++i) {
+      camera_ = cm->cameras()[i];
```
Comment (ericcurtin) [inline src/camera_session.cpp:34]: This doesn't account for the case where you for loop through everything and no camera was validated.

File: src/camera_session.cpp [hunk 11]
```diff
@@ -29,8 +29,19 @@ using namespace libcamera;
...
+      if (!CameraSession::validateConfig())
+        return;
+    }
+  } else if (opts.camera < static_cast<int>(cm->cameras().size())) {
     camera_ = cm->cameras()[opts.camera];
+    if (CameraSession::validateConfig()) {
+      EPRINT("Camera configuration invalid\n");
```
Comment (ericcurtin) [inline src/camera_session.cpp:41]: Please don't double log, this string would have already been printed in validateConfig

**Comment by ericcurtin** [comment]
Merged as a16d82cd6ae83fa53820336c611ee52c77620cf1 , had to handle all the error cases correctly, so rewrote a few bits

**Pull request closed without merging**